### PR TITLE
Feat: add switch widget

### PR DESCRIPTION
**Pull request by lp9052**
add switch widget

see #174

### CODE REVIEW

**Comment by lp9052** [comment]
Do we have permission to rerun failed test?

**Comment by maartenbreddels** [comment]
Thank you 🙏 

Apparently, you need to have write access for the repo to do that. I'm happy to restart it though

File: solara/components/switch.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+
+@solara.value_component(bool)
+def Switch(
+    *,
+    label=None,
+    value: Union[bool, solara.Reactive[bool]] = True,
+    on_value: Callable[[bool], None] = None,
+    disabled=False,
+    style: str = None,
```
Comment (maartenbreddels) [inline solara/components/switch.py:15]: See bc81d7d93a608438674118afb0f27183de956afd for how we prefer the style argument

Please also add classes.

File: solara/components/switch.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+     * `label`: The label to display next to the switch.
+     * `value`: The current value of the switch (True or False).
+     * `on_value`: A callback that is called when the switch is toggled.
+     * `disabled`: If True, the switch is disabled and cannot be used.
+     * `style`: A string of CSS styles to apply to the switch.
+    """
+    reactive_value = solara.use_reactive(value, on_value)
+    del value, on_value
+    children = []
```
Comment (maartenbreddels) [inline solara/components/switch.py:52]: See https://github.com/widgetti/solara/blob/6934ee5108ad2a05d10a50ae8b3ad16dcca22fb4/solara/components/button.py#L11 for how we can support any child in addition to a label argument

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Just a few minor points that would make this PR even more perfect 👍

**Comment by lp9052** [comment]
Thanks Maarten. 

I guess checkbox is one of the earlier piece that doesn't have all the new styles. All the enhancements make sense to me. I shall  update checkbox too. I'll start a new PR for checkbox.

**Comment by maartenbreddels** [comment]
> I guess checkbox is one of the earlier piece that doesn't have all the new styles. All the enhancements make sense to me. I shall update checkbox too. I'll start a new PR for checkbox.

indeed, and that would be great!

> one small concern, we are passing `[]` as default parameter, it's mutable.

Yes, that's also something I usually frown upon... we are using that in many places because it also keeps the typing simpler so lets stick to it for now and if we change it, we'll do it throughout the whole codebase.

**Comment by lp9052** [comment]
hmmm, the testing suit seems somewhat unstable. 

also I want to open up an issue about unit test doesn't all pass out of the box.

it requires some other packages as solar enterprise, etc.

**Comment by maartenbreddels** [comment]
Hmm, it's been a while since that test failed. Before we saw that one fail more often, see https://github.com/widgetti/solara/issues/131 for some comments/clues.
The integration test also has 1 flakey test.

**Comment by lp9052** [comment]
> Hmm, it's been a while since that test failed. Before we saw that one fail more often, see #131 for some comments/clues.
> The integration test also has 1 flakey test.

it's a quite interesting issue. I always like to see coverage info. And in my opinion 100% coverage is a must if we want to push this project further and accept more contributors

**Comment by maartenbreddels** [comment]
I agree. A possible option might be to explicitly set `intrusive_cancel` in https://solara.dev/api/use_thread and only test that intrusive_cancel feature separately. 
Still, I find it really suspicious/odd that even after 10 seconds it does not finish the thread execution. I've stared at this failure many times and thought about it many times, but it might need a fresh pair of eyes.

**Review by maartenbreddels — APPROVED** [review]
💯

**Comment by maartenbreddels** [comment]
There are 2 few missing I think
TODO:
 - [ ] Add to solara/website/pages/api/__init__.py
 - [ ] Add a gif 

I'm happy to pick this up if this is too much to ask. But this would also make this PR a nice template for new components.

**Comment by lp9052** [comment]
> There are 2 few missing I think
> 
> TODO:
> 
>  - [ ] Add to solara/website/pages/api/__init__.py
> 
>  - [ ] Add a gif to solara/website/public/api/
> 
> 
> 
> I'm happy to pick this up if this is too much to ask. But this would also make this PR a nice template for new components.

Give me a day or two. I might run into problem for gif. But I'll let you know. Thanks.

**Comment by lp9052** [comment]
Added these two pieces per suggestion.

**Comment by lp9052** [comment]
@maartenbreddels

**Comment by maartenbreddels** [comment]
Thank you! 🙏

**Pull request merged**